crates/starknet/src/starknet/mod.rs

```diff
@@ -705,6 +706,107 @@ impl Starknet {
         transaction_to_map.get_receipt()
     }
 
+    fn get_execute_call_info(
+        execution_info: &TransactionExecutionInfo,
+        address_to_class_hash_map: &HashMap<ContractAddress, ClassHash>,
+    ) -> DevnetResult<ExecutionInvocation> {
+        Ok(match &execution_info.execute_call_info {
+            Some(call_info) => match call_info.execution.failed {
+                false => ExecutionInvocation::Succeeded(FunctionInvocation::try_from_call_info(
+                    call_info,
+                    address_to_class_hash_map,
+                )?),
+                true => {
+                    ExecutionInvocation::Reverted(starknet_types::rpc::transactions::Reversion {
+                        revert_reason: execution_info
+                            .revert_error
+                            .clone()
+                            .unwrap_or("Revert reason not found".into()),
+                    })
+                }
+            },
+            None => {
+                match execution_info.revert_error.clone() {
+                    Some(revert_reason) => ExecutionInvocation::Reverted(
+                        starknet_types::rpc::transactions::Reversion { revert_reason },
+                    ),
+                    None => {
+                        return Err(Error::UnexpectedInternalError {
+                            msg: "Simulation contains neither call_info nor revert_error".into(),
+                        });
+                    }
+                }
+            }
+        })
+    }
+
+    fn get_call_info_invocation(
+        call_info_invocation: &Option<CallInfo>,
+        address_to_class_hash_map: &HashMap<ContractAddress, ClassHash>,
+    ) -> DevnetResult<Option<FunctionInvocation>> {
+        Ok(if let Some(call_info) = call_info_invocation {
+            Some(FunctionInvocation::try_from_call_info(call_info, address_to_class_hash_map)?)
+        } else {
+            None
+        })
+    }
+
+    pub fn get_transaction_trace_by_hash(
+        &self,
+        transaction_hash: TransactionHash,
+    ) -> DevnetResult<TransactionTrace> {
+        let transaction_to_map =
+            self.transactions.get(&transaction_hash).ok_or(Error::NoTransaction)?;
+
+        let mut state =
+            self.get_state_at(&BlockId::Tag(starknet_rs_core::types::BlockTag::Latest))?.clone();
+
+        let state_diff: ThinStateDiff = state.extract_state_diff_from_pending_state()?.into();
+        let state_diff =
+            if state_diff == ThinStateDiff::default() { None } else { Some(state_diff) };
```

Comment:
This is already the second time that we are introducing this hacky check.
It's been discussed in [this conversation](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/257/files#r1401714563) with @marioiordanov.

Maybe more logic could be extracted to a common function used by `get_transaction_trace` and `simulate_transactions`?

Reply:
fixed in 2fac26806808f07a966160c7249bb15458169ce3

Reply:
Ok not fixed and will not be fixed I guess...

I needed to revert that in f888865b0350b62903ce7d246c2437c806b39cee. The problem is that this refactor will cause extra clones of that state :(

Additionally, it gave me a headache (×_×;）